### PR TITLE
Optional dependecies

### PR DESCRIPTION
**Pull request by nvuletic**
This pull request includes:

- Addition of CMake options that allow the library user to not build bulky dependencies and extract only parts of this library. The options are `BUILD_PAHO_MQTT`, `BUILD_POCO` and `BUILD_GTEST`. If none are selected, then only the utilities is built, you can optionally take in `Poco` and `GTest`, and by enabling `Paho MQTT` you get the entire functionality of the `WolkSDK` with the `Wolk` class and its API.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -53,17 +57,58 @@ link_directories(${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
...
-target_include_directories(${PROJECT_NAME} PUBLIC "${CMAKE_CURRENT_LIST_DIR}/core")
+    add_library(${PROJECT_NAME} SHARED ${MQTT_SOURCE_FILES} ${CORE_HEADER_FILES} ${CONNECTIVITY_HEADER_FILES}
+            ${MODEL_HEADER_FILES} ${PERSISTENCE_HEADER_FILES} ${PROTOCOL_HEADER_FILES} ${SERVICE_HEADER_FILES})
+    target_link_libraries(${PROJECT_NAME} ${PROJECT_NAME}Utility paho-mqttpp3 paho-mqtt3as)
+    set_target_properties(${PROJECT_NAME} PROPERTIES INSTALL_RPATH "$ORIGIN")
+    add_dependencies(${PROJECT_NAME} ${PROJECT_NAME}Utility pahomqttc pahomqttcpp)
 
+    list(APPEND INSTALL_HEADER_FILES ${MQTT_HEADER_FILES})
+    list(APPEND INSTALL_TARGETS ${PROJECT_NAME})
+endif ()
```
Comment (srdjastankovic) [inline CMakeLists.txt:97]: endif what? 

endif(BUILD_PAHO_MQTT)

File: CMakeLists.txt [hunk 2]
```diff
@@ -53,17 +57,58 @@ link_directories(${CMAKE_LIBRARY_OUTPUT_DIRECTORY})
...
+# Create the install rule
+install(TARGETS ${INSTALL_TARGETS} LIBRARY DESTINATION ${CMAKE_INSTALL_PREFIX}/lib)
+if (BUILD_PAHO_MQTT)
+    install(FILES ${CORE_HEADER_FILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/include/wolksdk)
+    install(FILES ${CONNECTIVITY_HEADER_FILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/include/wolksdk/connectivity)
+    install(FILES ${MODEL_HEADER_FILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/include/wolksdk/model)
+    install(FILES ${PERSISTENCE_HEADER_FILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/include/wolksdk/persistence)
+    install(FILES ${PROTOCOL_HEADER_FILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/include/wolksdk/protocol)
+    install(FILES ${SERVICE_HEADER_FILES} DESTINATION ${CMAKE_INSTALL_PREFIX}/include/wolksdk/service)
+endif ()
```
Comment (srdjastankovic) [inline CMakeLists.txt:108]: endif what? 

endif(BUILD_PAHO_MQTT)

File: dependencies/CMakeLists.txt [hunk 3]
```diff
@@ -5,64 +5,69 @@ set(PAHO_MQTT_CPP_VERSION "1.1")
...
+            GIT_TAG "v${PAHO_MQTT_CPP_VERSION}"
+            UPDATE_COMMAND ""
+            CMAKE_ARGS -DPAHO_WITH_SSL=TRUE
+            -DPAHO_MQTT_C_INCLUDE_DIRS=${CMAKE_PREFIX_PATH}/include
+            -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}
+            -DCMAKE_INSTALL_PREFIX=${CMAKE_PREFIX_PATH}
+            -DCMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}
+            DEPENDS pahomqttc
+            )
+endif ()
```
Comment (srdjastankovic) [inline dependencies/CMakeLists.txt:30]: endif what? 

endif(BUILD_PAHO_MQTT)

File: dependencies/CMakeLists.txt [hunk 4]
```diff
@@ -5,64 +5,69 @@ set(PAHO_MQTT_CPP_VERSION "1.1")
...
-)
+if (BUILD_GTEST)
+    ExternalProject_Add(libgtest
+            GIT_REPOSITORY "https://github.com/google/googletest"
+            GIT_TAG "release-${GTEST_VERSION}"
+            CMAKE_ARGS -DBUILD_SHARED_LIBS=ON
+            -DCMAKE_INSTALL_PREFIX=${CMAKE_PREFIX_PATH}
+            -DCMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}
+            )
+endif ()
```
Comment (srdjastankovic) [inline dependencies/CMakeLists.txt:40]: endif what? 

endif(BUILD_PAHO_MQTT)

File: dependencies/CMakeLists.txt [hunk 5]
```diff
@@ -5,64 +5,69 @@ set(PAHO_MQTT_CPP_VERSION "1.1")
...
+            -DENABLE_SEVENZIP=OFF
+            -DENABLE_ZIP=OFF
+            -DENABLE_APACHECONNECTOR=OFF
+            -DENABLE_CPPPARSER=OFF
+            -DENABLE_POCODOC=OFF
+            -DENABLE_PAGECOMPILER=OFF
+            -DENABLE_PAGECOMPILER_FILE2PAGE=OFF
+            -DCMAKE_INSTALL_PREFIX=${CMAKE_PREFIX_PATH}
+            )
+endif ()
```
Comment (srdjastankovic) [inline dependencies/CMakeLists.txt:73]: endif what? 

endif(BUILD_PAHO_MQTT)

Comment (srdjastankovic) [reply dependencies/CMakeLists.txt:73]: Add new line at the end of the file

File: CMakeLists.txt (hunk 1 above)
Comment (nvuletic) [reply CMakeLists.txt:97]: The endif() condition is not necessary, it is only in CMake for backwards compatibility reasons. [Link to documentation](https://cmake.org/cmake/help/latest/command/endif.html).

Even next to that, all the if statements are only followed by endif, and not if, they are all 1-depth ifs, no-body can get lost in this.

Comment (srdjastankovic) [reply CMakeLists.txt:97]: If so, then remove this endif (NOT DEFINED CMAKE_LIBRARY_OUTPUT_DIRECTORY) with endif()

Comment (nvuletic) [reply CMakeLists.txt:97]: Done.

File: dependencies/CMakeLists.txt [hunk 6]
```diff
@@ -5,64 +5,69 @@ set(PAHO_MQTT_CPP_VERSION "1.1")
...
-ExternalProject_Add(pahomqttc
-    GIT_REPOSITORY "https://github.com/eclipse/paho.mqtt.c"
-    GIT_TAG "v${PAHO_MQTT_C_VERSION}"
-    UPDATE_COMMAND ""
-    CMAKE_ARGS -DPAHO_WITH_SSL=TRUE
-               -DCMAKE_LIBRARY_OUTPUT_DIRECTORY=${CMAKE_LIBRARY_OUTPUT_DIRECTORY}
-               -DCMAKE_INSTALL_PREFIX=${CMAKE_PREFIX_PATH}
-               -DCMAKE_PREFIX_PATH=${CMAKE_PREFIX_PATH}
-)
+if (BUILD_PAHO_MQTT)
```
Comment (milanzlicic) [inline dependencies/CMakeLists.txt:8]: I think we should name this connectivity or something similar

**Review by milanzlicic — CHANGES_REQUESTED** [review]

**Review by milanzlicic — APPROVED** [review]

**Pull request merged**